### PR TITLE
feat(workbench): add personal workbench module for permission management

### PR DESCRIPTION
**Pull request by Han-Ya-Jun**
## What changed

Add a new **Personal Workbench** module that provides a centralized view for gateway maintainers and applicants to manage their permission-related tasks.

### New Endpoints (6 total)

| Category | Endpoint | Description |
|----------|----------|-------------|
| 我的代办 | `GET /backend/me/workbench/permissions/gateway/pending/` | Pending gateway permission applications (for maintainers) |
| 我的代办 | `GET /backend/me/workbench/permissions/mcp/pending/` | Pending MCP Server permission applications |
| 我的申请 | `GET /backend/me/workbench/permissions/gateway/applied/` | My submitted gateway permission applications |
| 我的申请 | `GET /backend/me/workbench/permissions/mcp/applied/` | My submitted MCP Server permission applications |
| 我的已办 | `GET /backend/me/workbench/permissions/gateway/handled/` | Handled gateway permission records |
| 我的已办 | `GET /backend/me/workbench/permissions/mcp/handled/` | Handled MCP Server permission records |

### Key Design Decisions

- **RESTful URL structure**: `/me/workbench/permissions/{gateway,mcp}/{pending,applied,handled}/` with nested `include()` for clean separation
- **Mixin composition**: `WorkbenchPermissionMixin` provides shared `get_queryset()` logic for maintainer gateway filtering, composed with `generics.ListAPIView` via multiple inheritance
- **Precise maintainer matching**: Uses `GatewayHandler.list_gateways_by_user()` for exact permission matching
- **Standard patterns**: Uses `serializer_class` + `get_serializer()` consistent with DRF conventions
- **Input SLZ for docs only**: Query input serializers are for swagger documentation generation; actual filtering is handled by django-filter FilterSets

### Files Changed

- `apis/web/personal_workbench/` — views, serializers, filters, urls
- `tests/apis/web/personal_workbench/` — 29 test cases including authorization boundary and substring regression tests
- `urls.py` — route registration

### Test Coverage

- 29 tests all passing
- Includes authorization boundary tests (non-maintainer exclusion)
- Includes substring maintainer mismatch regression tests
- Covers all 6 endpoints with filter and edge case scenarios

### CODE REVIEW

**Review by wklken — COMMENTED** [review]
# PR #2704 Code Review 汇总报告

> 由 codex-internal (gpt-5.4) + Claude Opus 4.7 双模型 review，主 agent 汇总整理。

## 变更概述

- 新增 `personal_workbench` 模块，提供 6 个接口，覆盖 API 网关与 MCP Server 两类权限场景的“我的代办 / 我的申请 / 我的已办”。
- 抽出 `BaseWorkbenchListApi` 统一列表分页/序列化逻辑，并增加 `_get_user_maintainer_gateway_ids()` 处理 maintainer 精确匹配。
- 为权限申请/审批结果补充 ITSM 工单链接字段与相关文档、测试。

## 问题列表

### 🔴 Blocking

无。

### 🟠 Major

1. **`我的申请 - API 网关` 数据源设计有偏差，无法稳定覆盖已完成申请**  
   - `WorkbenchMyApplyGatewayPermissionListApi` 当前直接查询 `AppPermissionApply.objects.filter(applied_by=username)`。  
   - 但从现有审批流看，网关权限审批完成后，`AppPermissionApply` 记录并不是长期稳定的数据源；这会导致“我的申请”在真实流转后无法完整展示已通过/已驳回的历史。  
   - 现有测试通过手工构造 `approved` 状态的 `AppPermissionApply` 掩盖了这一点，未覆盖真实生命周期。  
   - 建议改为基于 `AppPermissionRecord` 出数，或统一抽象 pending + record 两类数据源。  
   - 来源：双方共同指出，Claude 更明确聚焦“生命周期/数据源错误”，Codex 也将其列为核心 major。[codex][claude]

2. **个人工作台的 maintainer 网关筛选未复用现有 tenant-aware 逻辑，存在多租户数据隔离风险**  
   - `_get_user_maintainer_gateway_ids()` 仅做 `_maintainers__contains=username` + `has_permission(username)` 精确过滤。  
   - 但仓库已有 `GatewayHandler.list_gateways_by_user(username, tenant_id)` 之类 tenant-aware 逻辑；当前实现未显式纳入 tenant 维度，多租户场景下可能把其他租户的待办一并带出来。  
   - 这属于权限隔离/数据可见性问题，建议直接复用现有 helper，避免逻辑漂移。  
   - Claude 另外补充，这个 helper 还有潜在的性能/可维护性问题（全字段拉取、Python 端过滤、重复维护），可以顺手用 `.only("id", "_maintainers")` 或直接复用现有 handler。 [codex][claude]

3. **`BaseWorkbenchListApi.list` 对分页的假设过强，未处理未分页分支**  
   - 当前实现直接对 `paginate_queryset()` 的结果做序列化并调用 `get_paginated_response()`。  
   - 一旦未来某个子类关闭分页或 paginator 返回 `None`，这里会直接 500。  
   - 建议直接删除覆写，复用 DRF 默认实现；或按官方逻辑补上 `if page is not None` 分支。  
   - 该问题主要由 Claude 提出，属于实现层面的稳定性缺陷。 [claude]

4. **MCP“已办/代办”语义和过滤条件不够统一，后续扩展容易漏掉终态或暴露不该展示的数据**  
   - `WorkbenchHandledMCPPermissionListApi` 用 `APPROVED/REJECTED` 白名单定义“已办”，而 gateway 侧是 `.exclude(status=PENDING)`。如果将来 MCP 增加 `CANCELLED` / `EXPIRED` 等终态，MCP 侧会自动漏掉。  
   - 同时 `WorkbenchPendingMCPPermissionListApi` 未显式说明/过滤“网关已下线”场景，可能把不应再处理的工单继续展示出来。  
   - 建议统一“已办”语义，并确认下线网关的工单是否应继续出现在工作台中。  
   - 这一组问题主要来自 Claude。 [claude]

### 🟡 Minor

1. **申请人展示直接返回原始 `applied_by`，与现有页面 display name 策略不一致**  
   - 现在多个 workbench serializer 都直接暴露原始用户名，多租户下前端展示可能退化、不易识别。  
   - 建议与现有权限页面保持一致，走 display name 转换。 [codex]

2. **`_get_user_maintainer_gateway_ids()` 对空用户名缺少短路保护**  
   - `_maintainers__contains=""` 会匹配所有网关，再交给 Python 端过滤；虽然正常认证流下问题不大，但补一个 `if not username: return []` 更稳妥。 [claude]

3. **MCP 过滤器能力略弱，可考虑补充 status 等字段筛选**  
   - 尤其 `我的申请 - MCP` 返回所有状态时，前端通常会希望按状态进一步筛选。 [claude]

4. **`permission_classes` 未显式声明，安全模型不够直观**  
   - 当前依赖全局默认配置可运行，但这是身份隔离敏感接口，显式写出认证/授权约束会更清晰。 [claude]

### 💬 Nit

- `WorkbenchMCPServerBaseSLZ.get_title()` 的 dict 分支目前看起来没有实际用到，可简化。[claude]
- `typing.List`、`# -*- coding: utf-8 -*-` 等风格细节可以顺手清理，但都不阻塞。[claude]
- filter / view 之间的命名和复用关系可以再补一点注释，提高可读性。[claude]

## 优点

- `_maintainers__contains` + `has_permission()` 的双层校验思路是对的，能够规避 `admin` 命中 `superadmin` 的子串误判，相关回归测试也补到了。[codex][claude]
- `personal_workbench` 模块的结构清晰：filters / serializers / urls / views 分层明确，`BaseWorkbenchListApi` 也体现了良好的抽象方向。[codex][claude]
- ITSM 工单链接能力接入比较完整，兼顾了模板已配置和回退场景，测试与 API 文档也同步补齐。[codex][claude]
- 测试覆盖整体较扎实，不仅覆盖了正常路径，也覆盖了子串匹配、空数据、越权边界等关键场景。[claude]

## 综合建议

整体实现方向没问题，功能完整度也不错，但**当前更适合在修完几个关键问题后再合并**：

1. 先修 `我的申请 - API 网关` 的数据源问题；
2. 把 maintainer 网关筛选切到 tenant-aware 的既有实现，避免多租户数据泄漏；
3. 处理 `BaseWorkbenchListApi.list` 的分页分支问题；
4. 顺手统一 MCP / gateway 对“已办”语义的定义，并补对应回归测试。

如果这几处补齐，这个 PR 的可维护性和线上稳健性会好很多。

---
*由 codex-internal (gpt-5.4) + Claude Opus 4.7 双模型 review，主 agent 汇总 | [from openclaw-internal]*

**Comment by Han-Ya-Jun** [comment]
## 📝 Review 反馈修复说明

已根据 review 意见逐项修复，提交: `62ecffc6d`

### 🟠 Major 修复

| # | 问题 | 修复方案 |
|---|------|---------|
| **M1** | `AppPermissionApply` 数据源是否稳定 | ✅ **无需修改** — 经代码考古确认：审批后 `status` 原地更新（`_handle_apply_status` → `save()`），记录不会删除/迁移，是稳定数据源 |
| **M2** | `_get_user_maintainer_gateway_ids` 未复用 tenant-aware 逻辑 | ✅ 改为复用 `GatewayHandler.list_gateways_by_user(username, tenant_id)`，内部已处理子串过滤 + 租户隔离 |
| **M3** | `BaseWorkbenchListApi.list` 未处理 `paginate_queryset` 返回 `None` | ✅ 删除 `list()` 覆写，完全复用 DRF `ListAPIView` 默认实现 |
| **M4** | MCP 与 Gateway "已办"语义不统一 | ✅ MCP 已办从 `status__in=[APPROVED, REJECTED]` 改为 `.exclude(status=PENDING)`，新增终态自动覆盖 |

### 🟡 Minor 修复

| # | 问题 | 修复方案 |
|---|------|---------|
| **m1** | `applied_by` 应走 display name 转换 | 暂不修改，当前其他模块也直接使用 username，后续统一处理 |
| **m2** | 空 username 缺短路保护 | ✅ 加了 `if not username: return []` |
| **m3** | MCP 过滤器缺 status 筛选 | ✅ `WorkbenchMCPPermissionApplyFilter` 增加 `status = ChoiceFilter(choices=...)` |
| **m4** | `permission_classes` 未显式声明 | ✅ `BaseWorkbenchListApi` 显式声明 `permission_classes = [IsAuthenticated]` |

### 💬 Nit 修复

- ✅ `typing.List` → `list`（Python 3.9+ 内置泛型）
- ✅ `get_title()` 移除不会触达的 `dict` 分支

### ✅ 测试验证

29 个测试全部通过，包括子串匹配回归测试和越权边界测试。

**Comment by Han-Ya-Jun** [comment]
## 📝 补充修复：m1 — applied_by display name 转换

经对比 `apis/web/permission/serializers.py` 中的 `AppPermissionApplyOutputSLZ`，确认原模块对 `applied_by` 做了 display name 转换处理：

```python
# 原 permission 模块的实现
def get_applied_by(self, obj):
    return ResourcePermissionHandler.convert_applied_by_to_display_name(
        obj.bk_app_code, obj.applied_by,
        self.context.get("gateway_tenant_mode"),
        self.context.get("gateway_tenant_id"),
    )
```

已在提交 `a0069a6d0` 中为个人工作台的 **全部 4 个输出序列化器** 补齐了该处理：

| Serializer | 数据来源 | 租户信息获取方式 |
|------------|---------|----------------|
| `WorkbenchGatewayPermissionApplyOutputSLZ` | `AppPermissionApply` | `obj.gateway.tenant_mode/tenant_id` |
| `WorkbenchGatewayPermissionRecordOutputSLZ` | `AppPermissionRecord` | `obj.gateway.tenant_mode/tenant_id` |
| `WorkbenchMCPPermissionApplyOutputSLZ` | `MCPServerAppPermissionApply` | `obj.mcp_server.gateway.tenant_mode/tenant_id` |
| `WorkbenchMCPPermissionHandledOutputSLZ` | `MCPServerAppPermissionApply` | `obj.mcp_server.gateway.tenant_mode/tenant_id` |

实现方式：将 `applied_by` 从 model 字段改为 `SerializerMethodField`，直接从关联的 gateway 对象读取 `tenant_mode` 和 `tenant_id`（无需 view 层传 context），与 `ResourcePermissionHandler.convert_applied_by_to_display_name()` 复用同一套逻辑。

29 个测试全部通过 ✅

File: src/dashboard/apigateway/apigateway/urls.py [hunk 1]
```diff
@@ -91,6 +91,8 @@
     ),
     # mcp server marketplace
     path("backend/mcp-marketplace/", include("apigateway.apis.web.mcp_marketplace.urls")),
+    # personal workbench
+    path("backend/personal-workbench/", include("apigateway.apis.web.personal_workbench.urls")),
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/urls.py]: restful 风格api 【未来个人中心除了工作台，而可能还有通知提醒之类的接口】

```
backend/me/workbench/
```

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/filters.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    applied_by = filters.CharFilter(lookup_expr="icontains")
+    status = filters.ChoiceFilter(choices=MCPServerAppPermissionApplyStatusEnum.get_choices())
+    keyword = filters.CharFilter(method="keyword_filter")
+
+    class Meta:
+        model = MCPServerAppPermissionApply
+        fields = ["bk_app_code", "applied_by", "status", "keyword"]
+
+    def keyword_filter(self, queryset, name, value):
+        return queryset.filter(Q(bk_app_code__icontains=value) | Q(mcp_server__name__icontains=value))
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/filters.py]: 需要补充mcp title

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            "applied_by",
+            "created_time",
+            "status",
+            "itsm_ticket_id",
+            "itsm_ticket_url",
+        ]
+        read_only_fields = fields
+
+    def get_gateway_name(self, obj) -> str:
+        return obj.gateway.name if obj.gateway else ""
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: 过度防御，这个表网关不可能为空？

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+    def get_expire_days_display(self, obj) -> str:
+        return PermissionApplyExpireDaysEnum.get_choice_label(obj.expire_days)
+
+    def get_grant_dimension_display(self, obj) -> str:
+        return GrantDimensionEnum.get_choice_label(obj.grant_dimension)
+
+    def get_applied_by(self, obj) -> str:
+        gateway = obj.gateway
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: 同上

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    WorkbenchGatewayPermissionApplyOutputSLZ,
+    WorkbenchGatewayPermissionQueryInputSLZ,
+    WorkbenchGatewayPermissionRecordOutputSLZ,
+    WorkbenchMCPPermissionApplyOutputSLZ,
+    WorkbenchMCPPermissionHandledOutputSLZ,
+    WorkbenchMCPPermissionQueryInputSLZ,
+)
+
+
+def _get_user_maintainer_gateway_ids(username: str, tenant_id: str = "") -> list[int]:
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: 不需要封装一个函数

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    1. _maintainers__contains 粗筛 + has_permission 精确过滤（避免子串误匹配）
+    2. tenant_id 维度的数据隔离（多租户场景）
+    """
+    if not username:
+        return []
+    gateways = GatewayHandler.list_gateways_by_user(username, tenant_id)
+    return [gw.id for gw in gateways]
+
+
+class BaseWorkbenchListApi(generics.ListAPIView):
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: 【重要】这里应该使用 Mixin， 不应该继承 generics.ListAPIView

Comment (wklken) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: class xxxPermissionMixin
   permission_classes = [IsAuthenticated]

File: src/dashboard/apigateway/apigateway/urls.py (hunk 1 above)
Comment (wklken) [reply src/dashboard/apigateway/apigateway/urls.py]: 使用`/me`， workbench 只是其中一个

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py (hunk 4 above)
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: 这个也是一定有的？ 不需要防御

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/urls.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    WorkbenchHandledGatewayPermissionListApi,
+    WorkbenchHandledMCPPermissionListApi,
+    WorkbenchMyApplyGatewayPermissionListApi,
+    WorkbenchMyApplyMCPPermissionListApi,
+    WorkbenchPendingGatewayPermissionListApi,
+    WorkbenchPendingMCPPermissionListApi,
+)
+
+urlpatterns = [
+    # 我的代办
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/urls.py]: - /me/workbench/permissions/{gateway, mcp}/{pending, applied, handled}/
- 使用include，不要平摊

Comment (wklken) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/urls.py]: 【重要】

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+
+def _get_user_maintainer_gateway_ids(username: str, tenant_id: str = "") -> list[int]:
+    """获取当前用户作为 maintainer 的所有网关 ID
+
+    复用 GatewayHandler.list_gateways_by_user，该方法内部已处理：
+    1. _maintainers__contains 粗筛 + has_permission 精确过滤（避免子串误匹配）
+    2. tenant_id 维度的数据隔离（多租户场景）
+    """
+    if not username:
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: 这里也是过度防御

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    PermissionApplyExpireDaysEnum,
+)
+from apigateway.apps.permission.models import AppPermissionApply, AppPermissionRecord
+from apigateway.biz.permission.permission import ResourcePermissionHandler
+from apigateway.service.bk_itsm import ItsmPermissionApplyHelper
+
+# ========== 查询输入序列化器 ==========
+
+
+class WorkbenchGatewayPermissionQueryInputSLZ(serializers.Serializer):
```
Comment (wklken) [inline src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py:36]: input slz 没有实际使用(仅作为文档生成，但是要同步修改确保一致性），实际生效的是filter

文档备注下

File: src/dashboard/apigateway/apigateway/urls.py (hunk 1 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/urls.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py (hunk 4 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py (hunk 3 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py (hunk 5 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py (hunk 6 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py (hunk 4 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/urls.py (hunk 7 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/urls.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py (hunk 9 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/serializers.py:36]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py (hunk 8 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/views.py]: done

File: src/dashboard/apigateway/apigateway/apis/web/personal_workbench/filters.py (hunk 2 above)
Comment (Han-Ya-Jun) [reply src/dashboard/apigateway/apigateway/apis/web/personal_workbench/filters.py]: done

**Comment by wklken** [comment]
review again

**Review by wklken — APPROVED** [review]
LGTM
